src/absil/ilread.fs

```diff
@@ -3861,69 +3864,142 @@ let ClosePdbReader pdb =
     | None -> ()
 #endif
 
-let OpenILModuleReader infile opts = 
-
-   try 
-        let mmap = MemoryMappedFile.Create infile
-        let modul, ilAssemblyRefs, pdb = genOpenBinaryReader infile mmap opts
-        { modul = modul 
-          ilAssemblyRefs=ilAssemblyRefs
-          dispose = (fun () -> 
-            mmap.Close()
-            ClosePdbReader pdb) }
-    with _ ->
-        let mc = ByteFile(infile |> FileSystem.ReadAllBytesShim)
-        let modul, ilAssemblyRefs, pdb = genOpenBinaryReader infile mc opts
-        { modul = modul 
-          ilAssemblyRefs = ilAssemblyRefs
-          dispose = (fun () -> 
-            ClosePdbReader pdb) }
+type ILReaderMetadataSnapshot = (obj * nativeint * int) 
+type ILReaderTryGetMetadataSnapshot = (* path: *) string * (* snapshotTimeStamp: *) System.DateTime -> ILReaderMetadataSnapshot option
 
+[<RequireQualifiedAccess>]
+type MetadataOnlyFlag = Yes | No
+
+[<RequireQualifiedAccess>]
+type ReduceMemoryFlag = Yes | No
+
+type ILReaderOptions =
+    { pdbPath: string option
+      ilGlobals: ILGlobals
+      reduceMemoryUsage: ReduceMemoryFlag
+      metadataOnly: MetadataOnlyFlag
+      tryGetMetadataSnapshot: ILReaderTryGetMetadataSnapshot }
+
+[<Sealed>]
+type ILModuleReader(ilModule: ILModuleDef, ilAssemblyRefs: Lazy<ILAssemblyRef list>, dispose: unit -> unit) =
+    member x.ILModuleDef = ilModule
+    member x.ILAssemblyRefs = ilAssemblyRefs.Force()
+    interface IDisposable with
+        member x.Dispose() = dispose()
+    
 // ++GLOBAL MUTABLE STATE (concurrency safe via locking)
 type ILModuleReaderCacheLockToken() = interface LockToken
-let ilModuleReaderCache = new AgedLookup<ILModuleReaderCacheLockToken, (string * System.DateTime * ILScopeRef * bool), ILModuleReader>(0, areSimilar=(fun (x, y) -> x = y))
+let ilModuleReaderCache = new AgedLookup<ILModuleReaderCacheLockToken, (string * System.DateTime * ILScopeRef * bool * ReduceMemoryFlag * MetadataOnlyFlag), ILModuleReader>(0, areSimilar=(fun (x, y) -> x = y))
 let ilModuleReaderCacheLock = Lock()
 
-let OpenILModuleReaderAfterReadingAllBytes infile opts = 
+let stableFileHeuristicApplies (filename:string) = 
+    try 
+       let directory = Path.GetDirectoryName(filename)
+       directory.Contains("Reference Assemblies") || directory.Contains("packages") || directory.Contains("lib/mono")
```

Comment:
Maybe include path separators ? To avoid any project named "foopackageme" to get considered stable :)

Also is there any consequences for paket users where `/packages/` dlls can change on paket install/update ?

Reply:
Yes path does not change when paket updates a lib. But a gazillion of other things like file size, file version, last modified date,....

Reply:
So in that sense it's not different than me maintaining a lib folder and replacing a dll in it

Reply:
Yes except that the dll in the lib folder will be loaded into memory because it won't benefit from the heuristic :)

I was just wondering what is the failure case, is it just a single build in FCS that warn/fail if it build while the packages are updating or if there is more long lived consequences.

Reply:
>I was just wondering what is the failure case, is it just a single build in FCS that warn/fail if it build while the packages are updating or if there is more long lived consequences.

The failure would normally be an exception surfaced as a intellisense error saying "please reload"